elliptic-curve/src/point/non_identity.rs

```diff
@@ -103,6 +107,81 @@ impl<P> AsRef<P> for NonIdentity<P> {
     }
 }
 
+impl<const N: usize, P> BatchNormalize<[Self; N]> for NonIdentity<P>
+where
+    P: Curve + BatchNormalize<[P; N], Output = [P::AffineRepr; N]>,
+{
+    type Output = [NonIdentity<P::AffineRepr>; N];
+
+    fn batch_normalize(points: &[Self; N]) -> [NonIdentity<P::AffineRepr>; N] {
+        // Ensure casting is safe.
+        // This always succeeds because `NonIdentity` is `repr(transparent)`.
+        debug_assert_eq!(size_of::<P>(), size_of::<NonIdentity<P>>());
+        debug_assert_eq!(align_of::<P>(), align_of::<NonIdentity<P>>());
+
+        #[allow(unsafe_code)]
+        // SAFETY: `NonIdentity` is `repr(transparent)`.
+        let points: &[P; N] = unsafe { &*points.as_ptr().cast() };
+        let affine_points = <P as BatchNormalize<_>>::batch_normalize(points);
+
+        // Ensure `array::map()` can be optimized to a `memcpy`.
+        debug_assert_eq!(
+            size_of::<P::AffineRepr>(),
+            size_of::<NonIdentity<P::AffineRepr>>()
+        );
+        debug_assert_eq!(
+            align_of::<P::AffineRepr>(),
+            align_of::<NonIdentity<P::AffineRepr>>()
+        );
+
+        affine_points.map(|point| NonIdentity { point })
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl<P> BatchNormalize<[Self]> for NonIdentity<P>
+where
+    P: Curve + BatchNormalize<[P], Output = Vec<P::AffineRepr>>,
+{
+    type Output = Vec<NonIdentity<P::AffineRepr>>;
+
+    fn batch_normalize(points: &[Self]) -> Vec<NonIdentity<P::AffineRepr>> {
+        // Ensure casting is safe.
+        // This always succeeds because `NonIdentity` is `repr(transparent)`.
+        debug_assert_eq!(size_of::<P>(), size_of::<NonIdentity<P>>());
+        debug_assert_eq!(align_of::<P>(), align_of::<NonIdentity<P>>());
+
+        #[allow(unsafe_code)]
+        // SAFETY: `NonIdentity` is `repr(transparent)`.
+        let points: &[P] =
+            unsafe { core::slice::from_raw_parts(points.as_ptr().cast(), points.len()) };
+        let mut affine_points = <P as BatchNormalize<_>>::batch_normalize(points);
+
+        // Ensure casting is safe.
+        // This always succeeds because `NonIdentity` is `repr(transparent)`.
+        debug_assert_eq!(
+            size_of::<P::AffineRepr>(),
+            size_of::<NonIdentity<P::AffineRepr>>()
+        );
+        debug_assert_eq!(
+            align_of::<P::AffineRepr>(),
+            align_of::<NonIdentity<P::AffineRepr>>()
+        );
+
+        // `Vec::into_raw_parts()` is not stable yet.
+        let ptr = affine_points.as_mut_ptr();
+        let length = affine_points.len();
+        let capacity = affine_points.capacity();
+        core::mem::forget(affine_points);
+
+        #[allow(unsafe_code)]
+        // SAFETY: `NonIdentity` is `repr(transparent)`.
+        unsafe {
+            Vec::from_raw_parts(ptr.cast(), length, capacity)
+        }
```

Comment:
Similarly here it seems like all of this could be replaced by casting `Vec<P::AffineRepr>` to `Vec<NonIdentity<P::AffineRepr>>`

Reply:
I have to say that I find pointer casting much more fickle and harder to verify then using `*::from_raw_parts()`.

In any case, it is done.

Reply:
Case in point: I just quickly tried `Vec::as_ptr()`, which fails, because its not a pointer to the `Vec`, but to the inner slice.

On a side note: I can't use `slice::as_ptr()` because it returns a thin pointer. But slices require a fat pointer for the length.

Reply:
Alternative suggestion: https://github.com/RustCrypto/traits/pull/1896#discussion_r2144993168